pyxem/signals/diffraction_vectors.py

```diff
@@ -215,50 +309,116 @@ def get_magnitude_histogram(self, bins, *args, **kwargs):
 
         return ghis
 
-    def get_unique_vectors(self,
-                           distance_threshold=0):
-        """Obtain the unique diffraction vectors.
+    def get_unique_vectors(self, distance_threshold=0.01,
+                           method='distance_comparison', min_samples=1,
+                           return_clusters=False):
+        """Obtain the unique diffraction vectors by a distance
+        comparison or by clustering using DBSCAN [1].
 
         Parameters
         ----------
         distance_threshold : float
-            The minimum distance between diffraction vectors for them to be
-            considered unique diffraction vectors.
+            The minimum distance between diffraction vectors for them to
+            be considered unique diffraction vectors.
+        method : string
+            The method to use to determine unique vectors. Valid methods
+            are 'distance_comparison' or 'DBSCAN'. 'distance_comparison'
+            iteratively checks the distance between two vectors to
+            determine if they should belong to the same unique vector,
+            and if so, the unique vector is iteratively updated to the
+            average value of the two. 'DBSCAN' relys on the DBSCAN [1]
+            clustering algorithm, and uses an Eucledian distance metric.
+        min_samples : int, optional
+            The minimum number of not identical vectors within one
+            cluster for it to be considered a core sample, i.e. to not
+            be considered noise. Only used for method='DBSCAN'.
+        return_clusters : bool, optional
+            If True (False is default), the DBSCAN clustering result is
+            returned. Only used for method='DBSCAN'.
+
+        References
+        ----------
+        [1] https://scikit-learn.org/stable/modules/generated/sklearn.
+            cluster.DBSCAN.html
 
         Returns
         -------
-        unique_vectors : DiffractionVectors
-            A DiffractionVectors object containing only the unique diffraction
-            vectors in the original object.
+        unique_peaks : DiffractionVectors
+            The unique diffraction vectors obtained by clustering.
+        clusters : DBSCAN
+            The results from the clustering, given as class DBSCAN.
+            Only returned if method='DBSCAN' and return_labels is True.
         """
-        if self.axes_manager.navigation_dimension == 2:
-            gvlist = np.array([self.data[0, 0][0]])
+        # Flatten the array of peaks to reach dimension (n, 2), where n
+        # is the number of peaks.
+        peaks = self.data
+        peaks_flat = self.data.flat
+        peaks_all = np.concatenate(np.array(list(map(
+            lambda x: peaks_flat[x].ravel(),
```

Comment:
I'm generally not a huge fan of lambda expressions, mostly because they usually make things harder to read/understand, which is a problem for multi-developer code.

This is not the worst example of lambda expressions, in fact it might be a good idea, but you I think you could do this with a more obvious structure like:

for x in self.data.flat.ravel():

I've noticed you use lambdas a lot in your code in general, is there a reason you like them? @pc494 do you have an opinion here?

In any case since you essentially don't reuse the variables peaks and peaks_flat - I think I'd consider replacing their use directly with self.data and self.data.flat respectively

Reply:
> 
> 
> I'm generally not a huge fan of lambda expressions, mostly because they usually make things harder to read/understand, which is a problem for multi-developer code.
> 
> This is not the worst example of lambda expression use, in fact it might be a good idea, but I think you could do this with a more obvious structure like:
> 
> for x in self.data.flat.ravel():
> 
> I've noticed you use lambdas a lot in your code in general, is there a reason you like them more than doing this kind of thing? @pc494 do you have an opinion here?
> 
> In any case since you essentially don't reuse the variables peaks and peaks_flat - I think I'd consider replacing their use directly with self.data and self.data.flat respectively

I find it easier to make code work with map and lambda and also find it more elegant and simplistic, but it might come down to habits. In this case, I'll see if I can make it work in another way, since self.data.flat is a numpy.flatiter..